### PR TITLE
[feat] Introduce Foursquare cloud provider

### PR DESCRIPTION
**Pull request by macrigiuseppe**
This PR introduces a new FSQ Studio provider to be able to store and fetch maps from [Foursquare Studio](https://studio.foursquare.com/home)

studio_provider

fetch_studio_map

### CODE REVIEW

File: docs/api-reference/cloud-providers/cloud-provider.md [hunk 1]
```diff
@@ -23,9 +23,9 @@ The default provider class
 
 **Parameters**
 
--   `props` **[object][27]** 
-    -   `props.name` **[string][28]** 
-    -   `props.displayName` **[string][28]** 
+-   `props` **[object][27]**
+    -   `props.name` **[string][28]**
```
Comment (ibgreen) [inline docs/api-reference/cloud-providers/cloud-provider.md:27]: What are those numbers in brackets? Look like link tags - in that case shouldn't those tags be defined below? Or maybe this is a different markdown dialect?

File: examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-icon.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import React, {Component} from 'react';
+import {Icons} from 'kepler.gl/components';
+import PropTypes from 'prop-types';
+
+const style = {
+  background: 'black'
+};
+export default class UnfoldedIcon extends Component {
```
Comment (ibgreen) [inline examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-icon.js:8]: FoursquareIcon? (Nit: If we stop using default exports we won't be making such mistakes)

File: examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-icon.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import React, {Component} from 'react';
```
Comment (ibgreen) [inline examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-icon.js:1]: Should we just call this `foursquare-icon.js`

File: examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-provider.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+const NAME = 'FSQ-Studio';
+const DISPLAY_NAME = 'Foursquare Studio';
+const APP_NAME = 'Kepler.gl';
+
+const STUDIO_PRIVATE_STORAGE_ENABLED = true;
+const STUDIO_AUTH_AUDIENCE = 'https://foursquare.com/api/';
+const STUDIO_AUTH_SCOPE = 'openid profile email';
+
+const STUDIO_KEPLER_GL_IMPORT_SOURCE = 'kepler.gl-raw';
```
Comment (ibgreen) [inline examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-provider.js:14]: What does kepler.gl-raw mean? Is it a URL path, a parameter value, etc? Maybe add comment before this constant,

File: examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-provider.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    description: map.info.description,
+    name: map.info.title,
+    importSource: STUDIO_KEPLER_GL_IMPORT_SOURCE,
+    latestState: {
+      data: map
+    }
+  };
+}
+
+export default class FsqStudioProvider extends Provider {
```
Comment (ibgreen) [inline examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-provider.js:51]: I vote for `FoursquareProvider`

File: examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-provider.js [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+import {Auth0Client} from '@auth0/auth0-spa-js';
+import {KEPLER_FORMAT} from '@kepler.gl/cloud-providers';
+
+const NAME = 'FSQ-Studio';
+const DISPLAY_NAME = 'Foursquare Studio';
+const APP_NAME = 'Kepler.gl';
+
+const STUDIO_PRIVATE_STORAGE_ENABLED = true;
+const STUDIO_AUTH_AUDIENCE = 'https://foursquare.com/api/';
+const STUDIO_AUTH_SCOPE = 'openid profile email';
```
Comment (ibgreen) [inline examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-provider.js:12]:
```suggestion
const FOURSQUARE_PRIVATE_STORAGE_ENABLED = true;
const FOURSQUARE_AUTH_AUDIENCE = 'https://foursquare.com/api/';
const FOURSQUARE_AUTH_SCOPE = 'openid profile email';
```

File: examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-provider.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        ...headers,
+        'Content-Type': 'image/png'
+      },
+      body: thumbnail
+    });
+
+    return map;
+  }
+
+  async listMaps() {
```
Comment (ibgreen) [inline examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-provider.js:127]: Defining explicit return types on these methods would be helpful for reader.

File: examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-provider.js [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    const response = await fetch(`${this.apiURL}/v1/maps/${mapId}`, {
+      method: 'GET',
+      headers
+    });
+
+    const map = await response.json();
+
+    return Promise.resolve({
+      map: extractMapFromStudioResponse(map),
+      format: KEPLER_FORMAT
```
Comment (ibgreen) [inline examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-provider.js:157]: `KEPLER_FORMAT` string could probably be made a little longer / more self descriptive. There are a lot of different types of "formats" being managed by kepler.gl

File: examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-provider.js [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+  getManagementUrl() {
+    return this._folderLink;
+  }
+
+  async getAccessToken() {
+    return this._auth0.getTokenSilently();
+  }
+
+  async getHeaders() {
```
Comment (ibgreen) [inline examples/demo-app/src/cloud-providers/fsq-studio/fsq-studio-provider.js:169]: Nit: These helpers could perhaps be marked `protected` or `private`, i.e. visually separate the public / interface API we are implementing and the helpers.

File: examples/demo-app/src/cloud-providers/index.js [hunk 10]
```diff
@@ -22,15 +22,30 @@ import {AUTH_TOKENS} from '../constants/default-settings';
...
+import FsqStudioProvider from './fsq-studio/fsq-studio-provider';
+
+const {
+  DROPBOX_CLIENT_ID,
+  CARTO_CLIENT_ID,
+  FSQ_STUDIO_CLIENT_ID,
+  FSQ_STUDIO_DOMAIN,
+  FSQ_STUDIO_API_URL,
+  FSQ_STUDIO_USER_MAPS_URL
+} = AUTH_TOKENS;
```
Comment (ibgreen) [inline examples/demo-app/src/cloud-providers/index.js:34]: This is no longer just AUTH_TOKENS. To make the code easier to follow, maybe consider renaming to `CLOUD_PROVIDERS_CONFIGURATION` or similar?

File: examples/demo-app/webpack.config.js [hunk 11]
```diff
@@ -69,7 +69,11 @@ const CONFIG = {
       'MapboxAccessToken',
       'DropboxClientId',
       'MapboxExportToken',
-      'CartoClientId'
+      'CartoClientId',
+      'FSQStudioClientId',
```
Comment (ibgreen) [inline examples/demo-app/webpack.config.js:73]: Again recommend just going with `Foursquare`.
```suggestion
      'FoursquareClientId',
```

File: src/cloud-providers/src/provider.ts [hunk 12]
```diff
@@ -133,10 +135,10 @@ export default class Provider {
   /**
    * This method is called to determine whether user already logged in to this provider
    * @public
-   * @returns true if a user already logged in
+   * @returns {Promise<string>} return the access token if a user already logged in
```
Comment (ibgreen) [inline src/cloud-providers/src/provider.ts:138]: If we add types in typescript we should not add them to the TSDoc.

File: src/cloud-providers/src/provider.ts [hunk 13]
```diff
@@ -253,8 +255,7 @@ export default class Provider {
    * }
    */
   async downloadMap(loadParams): Promise<{map: SavedMap; format: string}> {
-    // @ts-expect-error
-    return;
+    return Promise.reject('You must implement downloadMap');
```
Comment (ibgreen) [inline src/cloud-providers/src/provider.ts:258]: Nit: In my opinion it is cleaner to just throw an error and the async function will convert that to a rejected promise.

File: test/browser/components/modals/save-map-modal.spec.js [hunk 14]
```diff
@@ -1,5 +1,25 @@
+// Copyright (c) 2023 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
```
Comment (ibgreen) [inline test/browser/components/modals/save-map-modal.spec.js:3]: Do we needs to add this? 
- If we do, can't we just add  a one liner `// Copyright (c) 2023 Uber Technologies, Inc., MIT license`? That is what we do in other vis.gl frameworks.
- Should this be a FSQ copyright? Or is the year wrong? I assume Uber didn't do anything with this code in 2023.

**Review by ibgreen — APPROVED** [review]
Nicely done! Minor nits, feel free to ignore.

File: test/browser/components/modals/save-map-modal.spec.js (hunk 14 above)
Comment (macrigiuseppe) [reply test/browser/components/modals/save-map-modal.spec.js]: this is part of the open source code. only files in the example folder should have the FSQ copyright

File: docs/api-reference/cloud-providers/cloud-provider.md (hunk 1 above)
Comment (macrigiuseppe) [reply docs/api-reference/cloud-providers/cloud-provider.md:27]: the tags are defined at the end of the file https://github.com/keplergl/kepler.gl/blob/d975ea1e043a99039ffedf9771159ba1edfe63b4/docs/api-reference/cloud-providers/cloud-provider.md?plain=1#L209

File: examples/demo-app/webpack.config.js (hunk 11 above)
Comment (macrigiuseppe) [reply examples/demo-app/webpack.config.js:73]: yeah, already replaced in the first PR

File: examples/demo-app/src/cloud-providers/index.js (hunk 10 above)
Comment (macrigiuseppe) [reply examples/demo-app/src/cloud-providers/index.js:34]: yeah we can do that

File: src/cloud-providers/src/provider.ts (hunk 12 above)
Comment (macrigiuseppe) [reply src/cloud-providers/src/provider.ts:139]: yeah correct. left over from migrating code from a js source

File: src/cloud-providers/src/provider.ts (hunk 13 above)
Comment (macrigiuseppe) [reply src/cloud-providers/src/provider.ts:259]: yeah i can do that

**Pull request merged**